source/MRVoxels/MRWeightedPointsShell.cpp

```diff
@@ -55,15 +55,26 @@ FunctionVolume weightedMeshToDistanceFunctionVolume( const Mesh & mesh, const We
             const auto coord = Vector3f( pos ) + Vector3f::diagonal( 0.5f );
             const auto voxelCenter = params.vol.origin + mult( params.vol.voxelSize, coord );
             auto pd = findClosestWeightedMeshPoint( voxelCenter, mesh, params.dist );
-            if ( !( pd.dist >= params.dist.minDistance && pd.dist < params.dist.maxDistance ) )
+            if ( !( pd.weightedDistance() < params.dist.maxDistance ) )
                 return cQuietNan;
-            if ( params.signDistanceByNormal )
+            if ( params.signDistanceByNormal == WeightedPointsToDistanceVolumeParams::SignDistanceByNormal::Everywhere
+                    && pd.weightedDistance() >= params.dist.minDistance )
             {
-                assert( pd.dist >= 0 );
                 if ( dot( mesh.pseudonormal( pd.mtp ), voxelCenter - mesh.triPoint( pd.mtp ) ) < 0 )
-                    pd.dist = -pd.dist;
+                    pd.distance = -pd.distance;
```

Comment:
The main problem with this change, is that it is performed not inside `findClosestWeightedMeshPoint`, where we keep using unmodified value for bidirectional shell for finding the closest triangle.